javalanglib/src/main/scala/java/lang/Character.scala

```diff
@@ -462,9 +462,90 @@ object Character {
   //def getDirectionality(c: scala.Char): scala.Byte
 
   /* Conversions */
-  def toUpperCase(c: scala.Char): scala.Char = c.toString.toUpperCase().charAt(0)
-  def toLowerCase(c: scala.Char): scala.Char = c.toString.toLowerCase().charAt(0)
-  //def toTitleCase(c: scala.Char): scala.Char
+  def toUpperCase(ch: scala.Char): scala.Char = {
+    val upper = toUpperCase(ch.toInt)
+    if (upper > scala.Char.MaxValue) ch
+    else upper.toChar
+  }
+
+  def toLowerCase(ch: scala.Char): scala.Char = {
+    val lower = toLowerCase(ch.toInt)
+    if (lower > scala.Char.MaxValue) ch
+    else lower.toChar
+  }
+
+  def toUpperCase(codePoint: scala.Int): scala.Int = {
+    val upperChars = _String.fromCodePoint(codePoint).toUpperCase
+
+    if (upperChars.length > 2) {
+      codePoint
+    } else if (upperChars.length == 2) {
+      val high = upperChars.charAt(0)
+      val low = upperChars.charAt(1)
+
+      if (isSurrogatePair(high, low))
+        toCodePoint(upperChars.charAt(0), upperChars.charAt(1))
+      else
+        codePoint
+    } else {
+      upperChars.charAt(0).toInt
+    }
+  }
+
+  def toLowerCase(codePoint: scala.Int): scala.Int = {
+    val lowerChars = _String.fromCodePoint(codePoint).toLowerCase
+
+    if (lowerChars.length > 2) {
+      codePoint
+    } else if (lowerChars.length == 2) {
+      val high = lowerChars.charAt(0)
+      val low = lowerChars.charAt(1)
+
+      if (isSurrogatePair(high, low))
+        toCodePoint(lowerChars.charAt(0), lowerChars.charAt(1))
+      else
+        codePoint
+    } else {
+      lowerChars.charAt(0).toInt
+    }
+  }
+
+  def toTitleCase(ch: scala.Char): scala.Char = {
+    val title = toTitleCase(ch.toInt)
+    if (title > scala.Char.MaxValue) ch
+    else title.toChar
+  }
+
+/*
+def format(codePoint: Int): String = "0x%04x".format(codePoint)
+
+for(cp <- 0 until Character.MAX_CODE_POINT) {
+  val titleCaseCP: Int = Character.toTitleCase(cp)
+  val upperCaseCP: Int = Character.toUpperCase(cp)
+
+  if (titleCaseCP != upperCaseCP) {
+    println(s"    case ${format(cp)} => ${format(titleCaseCP)}")
+  }
+}
+*/
+  def toTitleCase(codePoint: scala.Int): scala.Int = {
+    codePoint match {
```

Comment:
@sjrd does adding a `@switch` here matter, and a better question is how can I verify this myself?